drivers/sensor/bmp180/bmp180.c

```diff
@@ -0,0 +1,321 @@
+	if (ret < 0)
+		return ret;
+	uncompensated_temperature = sys_be16_to_cpu(uncompensated_temperature);
+
+	ret = bmp180_reg_write(dev, BMP180_REG_MEAS_CTRL, (BMP180_PRESS_OVER_REGVALUE | BMP_START_CONVERSION | BMP_SELECT_PRESSURE));
+	if (ret < 0)
+		return ret;
+	
+	ret = bmp180_wait_until_ready(dev);
+	if (ret < 0)
+		return ret;
+
+	uint8_t uncompensated_pressure_buff[3];
+	ret = bmp180_reg_read(dev, BMP180_REG_OUT_MSB, uncompensated_pressure_buff, 3);
+	if (ret < 0)
+		return ret;
+	
+	uint32_t uncompensated_pressure = \
+		uncompensated_pressure_buff[0] << 16 |
+		uncompensated_pressure_buff[1] << 8  | 
+		uncompensated_pressure_buff[2];
+	uncompensated_pressure >>= (8 - BMP180_PRESS_OVER);
+
+	bmp180_compensate_temp(data, uncompensated_temperature);
+	bmp180_compensate_press(data, uncompensated_pressure);
+	bmp180_calculate_altitude(data);
+	return 0;
+}
+
+static int bmp180_channel_get(const struct device *dev, enum sensor_channel chan, struct sensor_value *val) {
+	struct bmp180_data *data = to_data(dev);
+
+	switch (chan) {
+		case SENSOR_CHAN_AMBIENT_TEMP:
+			val->val1 = data->comp_temp / 10;
+			val->val2 = data->comp_temp % 10 * 100000;
+			break;
+		case SENSOR_CHAN_PRESS:
+			val->val1 = data->comp_press;
+			val->val2 = 0;
+			break;
+		case SENSOR_CHAN_ALTITUDE:
+			val->val1 = data->altitude;
+			val->val2 = 0;
+			break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static const struct sensor_driver_api bmp180_api_funcs = {
+	.sample_fetch = bmp180_sample_fetch,
+	.channel_get = bmp180_channel_get,
+};
+
+static void bmp180_compensate_temp(struct bmp180_data *data, int32_t adc_temp) {
+	data->x1 = ((adc_temp - data->comp_data.AC6) * data->comp_data.AC5) >> 15;
+	data->x2 = (data->comp_data.MC << 11) / (data->x1 + data->comp_data.MD);
+	data->b5 = data->x1 + data->x2;
+	data->comp_temp = (data->b5 + 8) >> 4;
+}
+
+static void bmp180_compensate_press(struct bmp180_data *data, int32_t adc_press) {
+
+	data->b6 = data->b5 - 4000;
+	data->x1 = (data->comp_data.B2 * ((data->b6 * data->b6) >> 12)) >> 11;
+	data->x2 = (data->comp_data.AC2 * data->b6) >> 11;
+	data->x3 = data->x1 + data->x2;
+	data->b3 = (((data->comp_data.AC1 * 4 + data->x3) << BMP180_PRESS_OVER) + 2) >> 2;
+
+	data->x1 = (data->comp_data.AC3 * data->b6) >> 13;
+	data->x2 = (data->comp_data.B1 * ((data->b6 * data->b6) >> 12)) >> 16;
+	data->x3 = ((data->x1 + data->x2) + 2) >> 2;
+	data->b4 = (data->comp_data.AC4 * (unsigned long)(data->x3 + 32768UL)) >> 15;
+	data->b7 = ((unsigned long)adc_press - data->b3) * (50000UL >> BMP180_PRESS_OVER);
+
+	if(data->b7 < 0x80000000)
+		data->comp_press = (data->b7 * 2) / data->b4;
+	else
+		data->comp_press = (data->b7 / data->b4) * 2;
+	
+	data->x1 = (data->comp_press >> 8) * (data->comp_press >> 8);
+	data->x1 = (data->x1 * 3038) >> 16;
+	data->x2 = (-7357 * data->comp_press) >> 16;
+
+	data->comp_press = data->comp_press + ((data->x1 + data->x2 + 3791) >> 4);
```

Comment:
Please provide a reference for where this compensation algorithm came from

Reply:
> Please provide a reference for where this compensation algorithm came from

done